src/GraphQL/GraphQLBuilderExtensions.cs

```diff
@@ -0,0 +1,827 @@
+                options.ValidationRules = (options.ValidationRules ?? DocumentValidator.CoreRules).Append(rule);
+                if (useForCachedDocuments)
+                {
+                    options.CachedDocumentValidationRules = (options.CachedDocumentValidationRules ?? Enumerable.Empty<IValidationRule>()).Append(rule);
+                }
+            });
+            return builder;
+        }
+        #endregion
+
+        #region - AddMetrics -
+        /// <summary>
+        /// Registers <see cref="InstrumentFieldsMiddleware"/> within the dependency injection framework and
+        /// configures it to be installed within the schema.
+        /// When <paramref name="enable"/> is <see langword="true"/>, configures execution to set
+        /// <see cref="ExecutionOptions.EnableMetrics"/> to <see langword="true"/>; otherwise leaves it unchanged.
+        /// </summary>
+        public static IGraphQLBuilder AddMetrics(this IGraphQLBuilder builder, bool enable = true)
+        {
+            builder.AddMiddleware<InstrumentFieldsMiddleware>();
+            if (enable)
+                builder.ConfigureExecution(options => options.EnableMetrics = true);
+            return builder;
+        }
+
+        /// <summary>
+        /// Registers <see cref="InstrumentFieldsMiddleware"/> within the dependency injection framework and
+        /// configures it to be installed within the schema.
+        /// Configures execution to run <paramref name="enablePredicate"/> and when <see langword="true"/>, sets
+        /// <see cref="ExecutionOptions.EnableMetrics"/> to <see langword="true"/>; otherwise leaves it unchanged.
+        /// </summary>
+        public static IGraphQLBuilder AddMetrics(this IGraphQLBuilder builder, Func<ExecutionOptions, bool> enablePredicate)
+        {
+            if (enablePredicate == null)
+                throw new ArgumentNullException(nameof(enablePredicate));
+
+            builder.AddMiddleware<InstrumentFieldsMiddleware>();
+            builder.ConfigureExecution(options =>
+            {
+                if (enablePredicate(options))
+                {
+                    options.EnableMetrics = true;
+                }
+            });
+            return builder;
+        }
+
+        /// <summary>
+        /// Registers <see cref="InstrumentFieldsMiddleware"/> within the dependency injection framework and
+        /// configures it to be installed within the schema when <paramref name="installPredicate"/> returns <see langword="true"/>.
+        /// Configures execution to run <paramref name="enablePredicate"/> and when <see langword="true"/>, sets
+        /// <see cref="ExecutionOptions.EnableMetrics"/> to <see langword="true"/>; otherwise leaves it unchanged.
+        /// </summary>
+        public static IGraphQLBuilder AddMetrics(this IGraphQLBuilder builder, Func<ExecutionOptions, bool> enablePredicate, Func<IServiceProvider, ISchema, bool> installPredicate)
```

Comment:
Separate `install` / `enable` settings look difficult.

Reply:
1. We may just add middleware into DI. For any reason.
2. We may 1 + then install that middleware into some particular schema.
3. [for Metrics only] We may 2 + then _enable_/_disable_ metrics on some particular schema with already installed metrics middleware.

😮 Looks difficult and forces a desire to rewrite everything.

Reply:
I'm open to suggestions for the metrics.  I tried to both provide simplistic methods for general configuration, while also providing complex methods for better control.  But since I don't really use metrics at all, I'm not sure exactly what would be best.

Also, we don't have to provide builders for _every_ scenario.  People can still use AddMiddleware and ConfigureExecution and ConfigureSchema as desired if they need something custom.  Or they can add code directly in their middleware or schema constructor.  I'm okay with removing some of the AddMetrics overloads if they are extraneous and it makes logical sense to do so.  We may want to see how all these changes impact the server project before making a decision.  (I know I had to make some changes.)

One thing the builders don't allow for is being able to independently configure execution options and schema options for multiple schemas.  Multiple schemas are supported, but all builder settings will apply to both schemas.  This is by design so that it does not overly complicate the builder design.  I felt that if someone has such a complicated project that they need multiple schemas _with different configurations_, then they can write their own code (and most likely have already done so).  It's honestly not that complicated.

Side note: the existing `Action<ISchema, IServiceProvider>` registrations can already be tailored so that they only run for a particular schema with something like this:

```cs
public static IGraphQLBuilder ConfigureSchema<TSchema>(this IGraphQLBuilder builder, Action<ISchema, IServiceProvider> action)
    where TSchema : ISchema
{
    if (action == null)
        throw new ArgumentNullException(nameof(action));
    Action<ISchema, IServiceProvider> action2 = (schema, services) =>
    {
        if (schema is TSchema)
            action(schema, services);
    };
    return builder.Register<IConfigureSchema>(new ConfigureSchema(action2));
}
```

Similarly this could be done for ConfigureExecution - although it would be required that `options.Schema` is set before the delegate runs.

Reply:
I know. I have no suggestions now. Let's simply keep it in mind.